src/main/scala/org/scalameter/japi/JBench.scala

```diff
@@ -0,0 +1,280 @@
+package org.scalameter.japi
+
+import java.lang.annotation.Annotation
+import java.lang.reflect.Method
+import org.scalameter.BasePerformanceTest._
+import org.scalameter._
+import org.scalameter.japi.annotation._
+import org.scalameter.reporting.{HtmlReporter, RegressionReporter}
+import scala.collection.mutable
+import scala.language.reflectiveCalls
+import scala.util.{DynamicVariable, Try}
+
+
+/** Base class for all annotation based benchmarks. */
+abstract class JBench extends BasePerformanceTest with Serializable {
+
+  // This is a hack that allows to define generators
+  // and contexts as fields.
+  //
+  // We basically set `constructionStarted` variable
+  // and create another instance via reflection.
+  // That reflected instance has initialized fields,
+  // but hasn't got populated setup tree.
+  if (!JBench.constructionStarted.value) {
```

Comment:
I didn't yet get a chance to dive into the code below, but I assume that you are doing this to prevent initialization order problems related to the fields.

While this indeed solves the problem, it could kick us back if `JBench` constructor code does some very specific side-effects, such as opening database connectors or obtaining file handles, etc., and lead to some unexpected behavior.

I suggest that we instead build the setupzipper contents only once the `BasePerformanceTest#executeTests` method gets called.
You would have to add an empty method called `rebuildSetupZipper` that is called at the beginning of `executeTests`.
In `JBench`, we can then override this method to set the `setupzipper` value for us, long after the constructor call completed, and generator or context fields got set.